sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiSinkSchemaChangeIT.java

```diff
@@ -0,0 +1,515 @@
+
+    int distinctCount = Integer.parseInt((String) queryResponse.get("f0_"));
+    int rangeMin = Integer.parseInt((String) queryResponse.get("f1_"));
+    int rangeMax = Integer.parseInt((String) queryResponse.get("f2_"));
+
+    LOG.info("total distinct count = {}, min = {}, max = {}", distinctCount, rangeMin, rangeMax);
+
+    return (rangeMax - rangeMin + 1) == distinctCount && distinctCount == MAX_N;
+  }
+
+  private static boolean checkRowDuplication(String tableSpec)
+      throws IOException, InterruptedException {
+    TableRow queryResponse =
+        Iterables.getOnlyElement(
+            BQ_CLIENT.queryUnflattened(
+                String.format("SELECT COUNT(DISTINCT(id)), COUNT(id) FROM %s", tableSpec),
+                PROJECT,
+                true,
+                true));
+
+    int distinctCount = Integer.parseInt((String) queryResponse.get("f0_"));
+    int totalCount = Integer.parseInt((String) queryResponse.get("f1_"));
+
+    LOG.info("total distinct count = {}, total count = {}", distinctCount, totalCount);
+
+    return distinctCount == totalCount;
+  }
+
+  static class UpdateSchemaDoFn extends DoFn<KV<Integer, Long>, Long> {
+
+    private final String projectId;
+    private final String datasetId;
+    private final String tableId;
+
+    private final String schemaString;
+
+    private transient BigqueryClient bqClient;
+
+    private static final String MY_COUNTER = "myCounter";
+
+    @StateId(MY_COUNTER)
+    @SuppressWarnings("unused")
+    private final StateSpec<@org.jetbrains.annotations.NotNull ValueState<Integer>> counter;
+
+    public UpdateSchemaDoFn(
+        String projectId, String datasetId, String tableId, TableSchema schema) {
+      this.projectId = projectId;
+      this.datasetId = datasetId;
+      this.tableId = tableId;
+      this.schemaString = BigQueryHelpers.toJsonString(schema);
+      this.bqClient = null;
+      this.counter = StateSpecs.value();
+    }
+
+    private int getRowCount(String tableSpec) throws IOException, InterruptedException {
+      TableRow queryResponse =
+          Iterables.getOnlyElement(
+              bqClient.queryUnflattened(
+                  String.format("SELECT COUNT(*) FROM %s", tableSpec), PROJECT, true, true));
+      return Integer.parseInt((String) queryResponse.get("f0_"));
+    }
+
+    @Setup
+    public void setup() {
+      bqClient = new BigqueryClient("StorageApiSinkSchemaChangeIT_UpdateSchema");
+    }
+
+    @Teardown
+    public void tearDown() {
+      return;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c, @StateId(MY_COUNTER) ValueState<Integer> counter)
+        throws InterruptedException {
+      int current = firstNonNull(counter.read(), 0);
+      if (current == 0) {
+        int rowCount = 0;
+        try {
+          rowCount = this.getRowCount(this.projectId + "." + this.datasetId + "." + this.tableId);
+        } catch (Exception e) {
+          LOG.error(e.toString());
```

Comment:
P.S. maybe we should rethrow the error here instead of just logging it? if `getRowCount` starts failing for whatever reason, `rowCount` will stay at 0 and the schema will never be updated, making the tests obsolete (since the purpose is to test the schema update feature).